src/credentials/StaticGroupDataProvider.cpp

```diff
@@ -0,0 +1,487 @@
+
+namespace chip {
+namespace Credentials {
+namespace {
+
+static constexpr size_t GroupFabricsMax    = 4;
+static constexpr size_t EndpointEntriesMax = 16;
+static constexpr size_t StateEntriesMax    = 16;
+static constexpr size_t KeyEntriesMax      = 16;
+
+class StaticGroupsProvider : public GroupDataProvider
+{
+    struct EndpointEntry : public GroupMapping
+    {
+        bool in_use = false;
+    };
+
+    struct StateEntry : public GroupState
+    {
+        bool in_use = false;
+    };
+
+    struct KeysEntry : public KeySet
+    {
+        bool in_use = false;
+    };
+
+    struct Fabric
+    {
+        bool in_use = false;
+        chip::FabricIndex fabric_index;
+        EndpointEntry endpoints[EndpointEntriesMax];
+        StateEntry states[StateEntriesMax];
+        KeysEntry keys[KeyEntriesMax];
+    };
+
+    class EndpointIterator : public Iterator<GroupId>
+    {
+    public:
+        EndpointIterator(Fabric * fabric, chip::EndpointId endpoint) : mFabric(fabric), mEndpoint(endpoint) { GotoNext(); }
+        uint16_t Count() override
+        {
+            uint16_t count = 0;
+            for (uint16_t i = 0; this->mFabric && i < EndpointEntriesMax; ++i)
+            {
+                EndpointEntry & entry = this->mFabric->endpoints[i];
+                if (entry.in_use && entry.endpoint == this->mEndpoint)
+                {
+                    count++;
+                }
+            }
+            return count;
+        }
+        bool HasNext() override { return this->mIndex < EndpointEntriesMax; }
+        GroupId Next() override
+        {
+            GroupId group = this->mIndex < EndpointEntriesMax ? this->mFabric->endpoints[this->mIndex++].group : 0;
+            GotoNext();
+            return group;
+        }
+
+    private:
+        void GotoNext()
+        {
+            while (this->mFabric && this->mIndex < EndpointEntriesMax)
+            {
+                EndpointEntry & entry = this->mFabric->endpoints[this->mIndex];
+                if (entry.in_use && entry.endpoint == this->mEndpoint)
+                {
+                    break;
+                }
+                this->mIndex++;
+            }
+        }
+        Fabric * mFabric           = nullptr;
+        chip::EndpointId mEndpoint = 0;
+        uint16_t mIndex            = 0;
+    };
+
+    class StateIterator : public Iterator<GroupState *>
+    {
+    public:
+        StateIterator(Fabric * fabric) : mFabric(fabric) { GotoNext(); }
+        uint16_t Count() override
+        {
+            uint16_t count = 0;
+            for (uint16_t i = 0; this->mFabric && i < StateEntriesMax; ++i)
+            {
+                StateEntry & entry = this->mFabric->states[i];
+                if (entry.in_use)
+                {
+                    count++;
+                }
+            }
+            return count;
+        }
+        bool HasNext() override { return this->mIndex < EndpointEntriesMax; }
+        GroupState * Next() override
```

Comment:
Perhaps signature should be `const GroupState * Next() const` with `mutable uint16_t mIndex`

Reply:
I have changed the signatures. Since these functions are now Next() const, the call won't affect the inner state of the object, the increments are now perform on the HasNext() call.